storage: quiesce less aggressively

tbg opened the pull request:
PR #26911 added a mechanism which during quiescence decisions ignored
nodes which (at some recent point in time) seemed non-live.
Unfortunately, the corresponding mechanism introduced to unquiesce when
the node becomes live is quite racy, at least when tormented with the
wild timings we throw at it in unit testing.  The basic discrepancy is
that unquiescing is triggered by an explicit liveness event (i.e. the kv
record being overwritten), whereas a node may seem non-live for mere
reasons of timing inbetween. In effect this means that quiescence is
more aggressive than unquiescence.

I explored various venues to address this (going as far as introducing
an unquiesce message to the Raft scheduler and serializing unquiescence
through the store Raft ticker goroutine), but it's hopeless. It's an
equally idle diversion to try to adjust the test liveness durations.

So, in this commit, the liveness check becomes more lenient: even when a
node seems non-live, we count it as live if there's a healthy connection
established to it. This effectively deals with issues in tests but also
seems like a good addition for real world workloads: when the liveness
range breaks down, we don't want to aggressively quiesce.

Fixes #27545.
Fixes #27607.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/27617)
<!-- Reviewable:end -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27617#-:-LHZVMRB5Yj1WhNVuMj2:bnfp4nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/store.go, line 3799 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27617#-LHZUnFiBfWBvaokEbhc:-LHZUnFiBfWBvaokEbhd:bu7ahd7) ([raw file](https://github.com/cockroachdb/cockroach/blob/07324eb606316aa895c4d301b7976bea1d51cdec/pkg/storage/store.go#L3799)):*
> ```Go
> 			if s.cfg.NodeLiveness != nil {
> 				nextMap := s.cfg.NodeLiveness.GetIsLiveMap()
> 				for nodeID := range nextMap {
> ```

`for nodeID, isLive := range nextMap {`

is more clear.
___
*[pkg/storage/store.go, line 3804 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27617#-LHZUz50BkWvBiYVZNmA:-LHZUz50BkWvBiYVZNmB:b-re8e66) ([raw file](https://github.com/cockroachdb/cockroach/blob/07324eb606316aa895c4d301b7976bea1d51cdec/pkg/storage/store.go#L3804)):*
> ```Go
> 					}
> 					// Liveness claims that this node is down, but ConnHealth gets the last say
> 					// because we'd rather quiesce a range too few than one too many.
> ```

nit: "too little than one too often" sounds better. Your call
___
*[pkg/storage/store.go, line 3806 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27617#-LHZVPqLCdRhVl5xbns6:-LHZVPqLCdRhVl5xbns7:b-vr2d7q) ([raw file](https://github.com/cockroachdb/cockroach/blob/07324eb606316aa895c4d301b7976bea1d51cdec/pkg/storage/store.go#L3806)):*
> ```Go
> 					// because we'd rather quiesce a range too few than one too many.
> 					//
> 					// NB: This has false negatives. If a node doesn't have a conn open to it
> ```

This could also have false positives because of the delay between nodes going down and rpc heartbeats being sent, but that seems fine. The range will eventually quiesce. Might be worth mentioning this and further justification for why we're conservative here.
___
*[pkg/storage/store.go, line 3814 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27617#-LHZVFgJDSXsQk5OU2Yn:-LHZVFgJDSXsQk5OU2Yo:bb749ij) ([raw file](https://github.com/cockroachdb/cockroach/blob/07324eb606316aa895c4d301b7976bea1d51cdec/pkg/storage/store.go#L3814)):*
> ```Go
> 					nextMap[nodeID] = (s.cfg.NodeDialer.ConnHealth(nodeID) == nil)
> 				}
> 				log.Warningf(ctx, "TSX liveness map %v", nextMap)
> ```

TSX?


<!-- Sent from Reviewable.io -->

bdarnell reviewed (approved):
> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27617)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

tbg committed:
storage: quiesce less aggressively

PR #26911 added a mechanism which during quiescence decisions ignored
nodes which (at some recent point in time) seemed non-live.
Unfortunately, the corresponding mechanism introduced to unquiesce when
the node becomes live is quite racy, at least when tormented with the
wild timings we throw at it in unit testing.  The basic discrepancy is
that unquiescing is triggered by an explicit liveness event (i.e. the kv
record being overwritten), whereas a node may seem non-live for mere
reasons of timing inbetween. In effect this means that quiescence is
more aggressive than unquiescence.

I explored various venues to address this (going as far as introducing
an unquiesce message to the Raft scheduler and serializing unquiescence
through the store Raft ticker goroutine), but it's hopeless. It's an
equally idle diversion to try to adjust the test liveness durations.

So, in this commit, the liveness check becomes more lenient: even when a
node seems non-live, we count it as live if there's a healthy connection
established to it. This effectively deals with issues in tests but also
seems like a good addition for real world workloads: when the liveness
range breaks down, we don't want to aggressively quiesce.

Fixes #27545.
Fixes #27607.

Release note: None

tbg reviewed (commented):
(Sorry, hadn't submitted my comments before merging)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27617#-:-LHbo4B8CMIre3aSsdX2:b-xa5lm2)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/store.go, line 3799 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27617#-LHZUnFiBfWBvaokEbhc:-LHZj3DOBqpN3z_TbVFo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/07324eb606316aa895c4d301b7976bea1d51cdec/pkg/storage/store.go#L3799)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`for nodeID, isLive := range nextMap {`

is more clear.
</blockquote></details>

Done.
___
*[pkg/storage/store.go, line 3804 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27617#-LHZUz50BkWvBiYVZNmA:-LHZkg_sEsSiluG1hEeC:bg4tfst) ([raw file](https://github.com/cockroachdb/cockroach/blob/07324eb606316aa895c4d301b7976bea1d51cdec/pkg/storage/store.go#L3804)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: "too little than one too often" sounds better. Your call
</blockquote></details>

That doesn't sound right to me, but I'm a foreigner and probably just don't know better. Done.
___
*[pkg/storage/store.go, line 3806 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27617#-LHZVPqLCdRhVl5xbns6:-LHZkTF-4YYmsUJ27cb9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/07324eb606316aa895c4d301b7976bea1d51cdec/pkg/storage/store.go#L3806)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This could also have false positives because of the delay between nodes going down and rpc heartbeats being sent, but that seems fine. The range will eventually quiesce. Might be worth mentioning this and further justification for why we're conservative here.
</blockquote></details>

Done.
___
*[pkg/storage/store.go, line 3814 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27617#-LHZVFgJDSXsQk5OU2Yn:-LHZkTdkEO83WES47cCa:b-ggjkvv) ([raw file](https://github.com/cockroachdb/cockroach/blob/07324eb606316aa895c4d301b7976bea1d51cdec/pkg/storage/store.go#L3814)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

TSX?
</blockquote></details>

The thing I grep for when I printf debug. Removed.


<!-- Sent from Reviewable.io -->
